src/storage/tracked.rs

```diff
@@ -0,0 +1,378 @@
+    T: Component<Storage = TrackedStorage<T, S>> + Clone + Send + Sync,
+    D: Deref<Target = MaskedStorage<T>>,
+{
+    /// Returns a bitset with all inserted and modified components added.
+    /// This method is only provided if you're using `TrackedStorage`.
+    pub fn changed_tracked(&self) -> &BitSet {
+        self.data.inner.changed()
+    }
+
+    /// Returns an iterator over the change events generated by the `TrackedStorage`.
+    pub fn change_events_tracked(&self) -> ChangeEvents {
+        self.data.inner.change_events()
+    }
+}
+
+impl<'e, S, T, D> Storage<'e, T, D>
+where
+    S: UnprotectedStorage<T> + Send + Sync + 'static,
+    T: Component<Storage = TrackedStorage<T, S>> + Clone + Send + Sync,
+    D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Maintains the `TrackedStorage`.
+    ///
+    /// You can only call this in case your component implements `PartialEq`.
+    /// This will compare the cache with the current storage and generate change events
+    /// in case the `PartialEq` implementation says that two components are different.
+    ///
+    /// If you don't care about `Change::Modified` events, you don't have to call this method.
+    ///
+    /// ## When should I call this method?
+    ///
+    /// You should make sure that it gets called before you need the information which component
+    /// has been modified. E.g. in case you have several systems writing to the component, then
+    /// several reading from it, you can just call `maintain_tracked` once in between.
+    pub fn maintain_tracked(&mut self)
+    where
+        T: PartialEq,
+    {
+        let (set, inner) = self.data.open_mut();
+        unsafe {
+            inner.maintain(set);
+        }
+    }
+
+    /// Resets the tracked storage. This clears all change events and the `changed` bitset.
+    /// You most likely want to do this at the end of every frame.
+    pub fn reset_tracked(&mut self) {
+        let (_, inner) = self.data.open_mut();
+        unsafe {
+            inner.reset();
+        }
+    }
+}
+
+#[derive(Derivative)]
+#[derivative(Default(bound = "S: Default"))]
+pub struct TrackedStorage<C, S = DenseVecStorage<C>> {
+    /// All `Inserted` and `Changed` components are marked.
+    changed: BitSet,
+    changes: Vec<Change>,
+    _marker: PhantomData<C>,
+    old: S,
+    storage: S,
+}
+
+impl<C, S> TrackedStorage<C, S>
+where
+    C: Clone,
+    S: UnprotectedStorage<C>,
+{
+    /// Returns a reference to the `changed` bitset,
+    /// which contains all components that have been inserted or modified
+    /// since the last `reset`.
+    pub fn changed(&self) -> &BitSet {
+        &self.changed
+    }
+
+    pub fn change_events<'a>(&'a self) -> ChangeEvents<'a> {
+        let inner = self.changes.iter().enumerate();
+
+        ChangeEvents { inner }
+    }
+
+    unsafe fn reset(&mut self) {
+        for id in &self.changed {
+            let elem = self.old.get_mut(id);
+            *elem = self.storage.get(id).clone();
```

Comment:
For some reason this I had hard time understanding this code the first time I was reading it. Propably because there is `let` instead of `let mut` here. Dunno how to make it better though.

Reply:
Me too, I think we're both used to the mutable binding because of ECS resources which require mutability for `DerefMut`.